File: src/lib/wow/archive/casc/build-cache.ts
/**
 * Build cache for CDN-downloaded CASC data, ported from wow.export
 * (src/js/casc/build-cache.js). UI cache-size tracking removed.
 */
import { promises as fsp } from 'fs';
import path from 'path';

import { write } from '@/lib/wow/log';

import { BufferWrapper } from '../../formats/buffer';
import { constants } from '../../formats/constants';
import { createDirectory, deleteDirectory, readJSON } from '../../formats/generics';
import { wowConfig } from '../../server/config';

let cacheIntegrity: Record<string, string> | null = null;
let integrityInitPromise: Promise<void> | null = null;
let integritySaveChain: Promise<void> = Promise.resolve();

/** Initialize the cache integrity system (lazy, once). */
async function ensureCacheIntegrity(): Promise<Record<string, string>> {
  if (cacheIntegrity) return cacheIntegrity;
  if (!integrityInitPromise) {
    integrityInitPromise = (async () => {
      try {
        const integrity = await readJSON(constants.CACHE.INTEGRITY_FILE, false);
        if (integrity === null) {
          throw new Error(`File cannot be accessed or contains malformed JSON: ${constants.CACHE.INTEGRITY_FILE}`);
        }
        cacheIntegrity = integrity as Record<string, string>;
      } catch (e) {
        write('Unable to load cache integrity file; entire cache will be invalidated.');
        write((e as Error).message);
        cacheIntegrity = {};
      }
    })();
  }
  await integrityInitPromise;
  return cacheIntegrity!;
}

export class BuildCache {
  key: string;

  meta: { lastAccess?: number } = {};

  cacheDir: string;

  manifestPath: string;

  constructor(key: string) {
    this.key = key;
    this.cacheDir = path.join(constants.CACHE.DIR_BUILDS, key);
    this.manifestPath = path.join(this.cacheDir, constants.CACHE.BUILD_MANIFEST);
  }

  /** Initialize the build cache instance. */
  async init(): Promise<void> {
    // Create cache directory if needed.
    await fsp.mkdir(this.cacheDir, { recursive: true });

    // Load manifest values.
    try {
      const manifest = JSON.parse(await fsp.readFile(this.manifestPath, 'utf8'));
      Object.assign(this.meta, manifest);
    } catch (e) {
      write('No cache manifest found for %s', this.key);
    }

    // Save access update without blocking.
    this.meta.lastAccess = Date.now();
    void this.saveManifest();
  }

  /**
   * Attempt to get a file from this build cache.
   * Returns null if the file is not cached.
   * @param file File path relative to build cache.
   * @param dir Optional override directory.
   */
  async getFile(file: string, dir?: string): Promise<BufferWrapper | null> {
    try {
      const filePath = this.getFilePath(file, dir);

      const integrity = await ensureCacheIntegrity();
      const integrityHash = integrity[filePath];

      // File integrity cannot be verified, reject.
      if (typeof integrityHash !== 'string') {
        write('Cannot verify integrity of file, rejecting cache (%s)', filePath);
        return null;
      }

      const data = await BufferWrapper.readFile(filePath);
      const dataHash = data.calculateHash('sha1', 'hex');

      // Reject cache if hash does not match.
      if (dataHash !== integrityHash) {
        write('Bad integrity for file %s, rejecting cache (%s != %s)', filePath, dataHash, integrityHash);
        return null;
      }

      return data;
    } catch (e) {
      return null;
    }
  }

  /** Get a direct path to a cached file. */
  getFilePath(file: string, dir?: string): string {
    return path.join(dir || this.cacheDir, file);
  }

  /**
   * Store a file in this build cache.
   * @param file File path relative to build cache.
   * @param data Data to store in the file.
   * @param dir Optional override directory.
   */
  async storeFile(file: string, data: BufferWrapper, dir?: string): Promise<void> {
    if (!(data instanceof BufferWrapper)) throw new Error('Data provided to cache.storeFile() must be of BufferWrapper type.');

    const filePath = this.getFilePath(file, dir);
    if (dir) await createDirectory(path.dirname(filePath));

    const integrity = await ensureCacheIntegrity();

    // Integrity checking.
    const hash = data.calculateHash('sha1', 'hex');
    integrity[filePath] = hash;

    await fsp.writeFile(filePath, data.raw);

    await this.saveCacheIntegrity();
  }

  /** Save the cache integrity to disk (serialized to avoid concurrent write races). */
  async saveCacheIntegrity(): Promise<void> {
    await createDirectory(path.dirname(constants.CACHE.INTEGRITY_FILE));
    const snapshot = { ...cacheIntegrity };
    integritySaveChain = integritySaveChain.then(() => fsp.writeFile(
      constants.CACHE.INTEGRITY_FILE,
      JSON.stringify(snapshot),
      'utf8',
    ));
    await integritySaveChain;
  }

  /** Save the manifest for this build cache. */
  async saveManifest(): Promise<void> {
    await fsp.writeFile(this.manifestPath, JSON.stringify(this.meta), 'utf8');
  }
}

/**
 * Run clean-up for stale build caches (mirrors wow.export's
 * 'casc-source-changed' clean-up hook; call after a CASC source is selected).
 */
export async function runCacheCleanup(): Promise<void> {
  let cacheExpire = Number(wowConfig.cacheExpiry) || 0;
  cacheExpire *= 24 * 60 * 60 * 1000;

  // If user sets cacheExpiry to 0 in the configuration, we completely
  // skip the clean-up process.
  if (cacheExpire === 0) {
    write('WARNING: Cache clean-up has been skipped due to cacheExpiry being %d', cacheExpire);
    return;
  }

  write('Running clean-up for stale build caches...');
  let entries;
  try {
    entries = await fsp.readdir(constants.CACHE.DIR_BUILDS, { withFileTypes: true });
  } catch (e) {
    return; // No cache directory yet.
  }
  const ts = Date.now();

  for (const entry of entries) {
    // We only care about directories with MD5 names.
    if (!entry.isDirectory() || entry.name.length !== 32) continue;

    let deleteEntry = false;
    const entryDir = path.join(constants.CACHE.DIR_BUILDS, entry.name);
    const entryManifest = path.join(entryDir, constants.CACHE.BUILD_MANIFEST);

    try {
      const manifestRaw = await fsp.readFile(entryManifest, 'utf8');
      const manifest = JSON.parse(manifestRaw) as { lastAccess?: number };

      if (manifest.lastAccess !== undefined && !Number.isNaN(manifest.lastAccess)) {
        const delta = ts - manifest.lastAccess;
        if (delta > cacheExpire) {
          deleteEntry = true;
          write('Build cache %s has expired (%d), marking for deletion.', entry.name, delta);
        }
      } else {
        deleteEntry = true;
        write('Unable to read lastAccess from %s, marking for deletion.', entry.name);
      }
    } catch (e) {
      // Manifest is missing or malformed.
      deleteEntry = true;
      write('Unable to read manifest for %s, marking for deletion.', entry.name);
    }

    if (deleteEntry) await deleteDirectory(entryDir);
  }
}

export default BuildCache;

File: golang/internal/wow/archive/casc/build_cache.go
package casc

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var (
	cacheIntegrity    map[string]string
	integrityInitOnce sync.Once
	integrityInitErr  error
	integrityMu       sync.RWMutex
)

func ensureCacheIntegrity() (map[string]string, error) {
	integrityInitOnce.Do(func() {
		cacheIntegrity, integrityInitErr = formats.ReadJSON(constants.Cache.IntegrityFile, false)
		if integrityInitErr != nil || cacheIntegrity == nil {
			log.Write("Unable to load cache integrity file; entire cache will be invalidated.")
			if integrityInitErr != nil {
				log.Write("%s", integrityInitErr.Error())
			}
			cacheIntegrity = map[string]string{}
			integrityInitErr = nil
		}
	})
	return cacheIntegrity, integrityInitErr
}

// BuildCacheMeta holds build cache manifest metadata.
type BuildCacheMeta struct {
	LastAccess int64 `json:"lastAccess,omitempty"`
}

// BuildCache caches CDN-downloaded CASC data per build.
type BuildCache struct {
	Key          string
	Meta         BuildCacheMeta
	CacheDir     string
	ManifestPath string
}

// NewBuildCache creates a build cache for the given key.
func NewBuildCache(key string) *BuildCache {
	cacheDir := filepath.Join(constants.Cache.DirBuilds, key)
	return &BuildCache{
		Key:          key,
		CacheDir:     cacheDir,
		ManifestPath: filepath.Join(cacheDir, constants.Cache.BuildManifest),
	}
}

// Init initializes the build cache instance.
func (c *BuildCache) Init() error {
	if err := os.MkdirAll(c.CacheDir, 0o755); err != nil {
		return err
	}
	if data, err := os.ReadFile(c.ManifestPath); err == nil {
		_ = json.Unmarshal(data, &c.Meta)
	} else {
		log.Write("No cache manifest found for %s", c.Key)
	}
	c.Meta.LastAccess = time.Now().UnixMilli()
	go c.SaveManifest()
	return nil
}

// GetFile attempts to get a file from this build cache.
func (c *BuildCache) GetFile(file string, dir string) (*buffer.Buffer, error) {
	filePath := c.GetFilePath(file, dir)
	if _, err := os.Stat(filePath); err != nil {
		return nil, nil
	}
	integrity, err := ensureCacheIntegrity()
	if err != nil {
		return nil, nil
	}
	integrityMu.RLock()
	integrityHash, ok := integrity[filePath]
	integrityMu.RUnlock()
	if !ok {
		log.Write("Cache file has no integrity record, ignoring (%s)", filePath)
		return nil, nil
	}
	data, err := buffer.ReadFile(filePath)
	if err != nil {
		return nil, nil
	}
	dataHash := data.CalculateHash("sha1", "hex")
	if dataHash != integrityHash {
		log.Write("Bad integrity for file %s, rejecting cache (%s != %s)", filePath, dataHash, integrityHash)
		return nil, nil
	}
	return data, nil
}

// GetFilePath returns a direct path to a cached file.
func (c *BuildCache) GetFilePath(file string, dir string) string {
	if dir == "" {
		dir = c.CacheDir
	}
	return filepath.Join(dir, file)
}

// StoreFile stores a file in this build cache.
func (c *BuildCache) StoreFile(file string, data *buffer.Buffer, dir string) error {
	filePath := c.GetFilePath(file, dir)
	if dir != "" {
		if err := formats.CreateDirectory(filepath.Dir(filePath)); err != nil {
			return err
		}
	}
	integrity, err := ensureCacheIntegrity()
	if err != nil {
		return err
	}
	hash := data.CalculateHash("sha1", "hex")
	integrityMu.Lock()
	integrity[filePath] = hash
	integrityMu.Unlock()
	if err := os.WriteFile(filePath, data.Raw(), 0o644); err != nil {
		return err
	}
	return c.SaveCacheIntegrity()
}

// SaveCacheIntegrity saves cache integrity to disk.
func (c *BuildCache) SaveCacheIntegrity() error {
	integrityMu.Lock()
	snapshot := make(map[string]string, len(cacheIntegrity))
	for k, v := range cacheIntegrity {
		snapshot[k] = v
	}
	integrityMu.Unlock()
	if err := formats.CreateDirectory(filepath.Dir(constants.Cache.IntegrityFile)); err != nil {
		return err
	}
	data, err := json.Marshal(snapshot)
	if err != nil {
		return err
	}
	return os.WriteFile(constants.Cache.IntegrityFile, data, 0o644)
}

// SaveManifest saves the manifest for this build cache.
func (c *BuildCache) SaveManifest() error {
	data, err := json.Marshal(c.Meta)
	if err != nil {
		return err
	}
	return os.WriteFile(c.ManifestPath, data, 0o644)
}
